src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -1598,26 +1594,52 @@
    DNS alt names, and several different fingerprint hashes of the certificate."
   [{:keys [csrdir signeddir cacert cacrl cakey]} :- CaSettings
    subject :- schema/Str]
-  (let [crl (utils/pem->ca-crl cacrl (utils/pem->ca-cert cacert cakey))]
-    (get-certificate-status* signeddir csrdir crl subject)))
+  (let [crl      (utils/pem->ca-crl cacrl (utils/pem->ca-cert cacert cakey))
+        is-cert? (fs/exists? (path-to-cert signeddir subject))]
+    (if is-cert?
+      (let [cert (utils/pem->cert (path-to-cert signeddir subject))]
+        (get-certificate-status* crl is-cert? subject cert))
+      (let [csr (utils/pem->csr (path-to-cert-request csrdir subject))]
+        (get-certificate-status* crl is-cert? subject csr)))))
+
+(schema/defn ^:always-validate get-CSRs-statuses :- [CertificateStatusResult]
+  "Get the status of only certificate requests by only going
+   through csrdir."
+  [csrdir :- schema/Str
+   crl :- CertificateRevocationList]
+  (let [pem-pattern   #"^.+\.pem$"
+        all-subjects  (map #(fs/base-name % ".pem") (fs/find-files csrdir pem-pattern))
+        all-CSRs      (map #(utils/pem->csr (path-to-cert-request csrdir %)) all-subjects)
+        is-cert?      false]
+    (map (partial get-certificate-status* crl is-cert?) all-subjects all-CSRs)))
+
+(schema/defn ^:always-validate get-signed-or-revoked-statuses :- [CertificateStatusResult]
```

Comment:
I guess that would also imply that we should change the names of `get-certificate-status*` and `get-certificate-status` as well. Open to other opinions on this, but just thinking it might be clearer to not use 'certificate' to mean 'cert or CSR'.

Reply:
Anyway, I think there's some simplification to be done around the `get-CSRs-statuses` and `get-signed-or-revoked-statuses` functions. They're nearly identical, other than one using the `csrdir` and one using the `signeddir`. Wondering if it makes sense for that directory to just be an argument to something.. Trying not to just do it for you haha so let me know if you want to chat more about what I had in mind.

Reply:
Yeah I split them up into two separate function mostly because I wanted to be clear. The other approach is just having one function basically take either `csrdir` or `signeddir` and a boolean that indicate whether we're dealing with a cert of csr and make the function call to `get-certificate-status*` from there. I do see the repetitiveness but thought that I wanted to be clear.